yt_dlp/extractor/cbc.py

```diff
@@ -209,289 +198,239 @@ def _real_extract(self, url):
+        'url': 'https://gem.cbc.ca/media/schitts-creek/s01e01',
+        'md5': '297a9600f554f2258aed01514226a697',
+        'info_dict': {
+            'id': 'schitts-creek/s01e01',
+            'ext': 'mp4',
+            'title': 'The Cup Runneth Over',
+            'description': 'md5:9bca14ea49ab808097530eb05a29e797',
+            'thumbnail': 'https://images.radio-canada.ca/v1/synps-cbc/episode/perso/cbc_schitts_creek_season_01e01_thumbnail_v01.jpg?im=Resize=(Size)',
+            'series': 'Schitt\'s Creek',
+            'season_number': 1,
+            'season': 'Season 1',
+            'episode_number': 1,
+            'episode': 'The Cup Runneth Over',
+            'episode_id': 'schitts-creek/s01e01',
+            'duration': 1309,
+            'categories': ['comedy'],
+        },
+        'params': {'format': 'bv'},
+        'skip': 'Geo-restricted to Canada',
+    }]
+    _API_BASE = 'https://services.radio-canada.ca/ott/cbc-api/v2/assets/'
 
     def _real_extract(self, url):
         video_id = self._match_id(url)
-        result = self._call_api(url, video_id)
-
-        m3u8_url = xpath_text(result, 'url', fatal=True)
-        formats = self._extract_m3u8_formats(re.sub(r'/([^/]+)/[^/?]+\.m3u8', r'/\1/\1.m3u8', m3u8_url), video_id, 'mp4', fatal=False)
-        if len(formats) < 2:
-            formats = self._extract_m3u8_formats(m3u8_url, video_id, 'mp4')
-        for f in formats:
-            format_id = f.get('format_id')
-            if format_id.startswith('AAC'):
-                f['acodec'] = 'aac'
-            elif format_id.startswith('AC3'):
-                f['acodec'] = 'ac-3'
+        video_info = self._download_json(self._API_BASE + video_id, video_id)
+        m3u8_info = self._download_json(video_info['playSession']['url'], video_id)
+
+        if m3u8_info.get('errorCode') == 1:
+            # errorCode 1 means geo-blocked
+            self.raise_geo_restricted(countries=['CA'])
+        elif m3u8_info.get('errorCode') not in [35, 0, None]:
+            # 35 means media unavailable, which is handled below. 0 means no
+            # error. None is there just in case. Any other errorCode can't be
+            # handled as we don't know what it is.
+            raise ExtractorError(
+                f'CBCGem said {m3u8_info.get("errorCode")} - {m3u8_info.get("message")}'
+            )
+
+        m3u8_url = m3u8_info.get('url')
+
+        # Sometimes the m3u8 URL is a not available, and errorCode is 35.
+        # This might only happen with member content. But just retrying seems
+        # to work fine.
+        max_retries = self.get_param('extractor_retries', 15)
+        i = 0
+        while m3u8_url is None:
+            if i == max_retries:
+                # Didn't work after all tries, give up
+                raise ExtractorError("Couldn't retrieve m3u8 URL")
+            m3u8_url = self._download_json(
+                video_info['playSession']['url'], video_id).get('url')
+            i += 1
```

Comment:
(untested)
```suggestion
        last_error, attempt, retries = None, -1, self.get_param('extractor_retries', 3)
        while attempt < retries:
            attempt += 1
            if last_error:
                self.report_warning('%s. Retrying ...' % last_error)
            m3u8_info = self._download_json(
                video_info['playSession']['url'], video_id,
                note='Downloading json metadata%s' % f' (attempt {attempt})')
            m3u8_url = m3u8_info.get('url')
            if m3u8_url:
                break
            elif m3u8_info.get('errorCode') == 1:
                self.raise_geo_restricted(countries=['CA'])
            else:
                last_error = f'{self.IE_NAME} said: {m3u8_info.get("errorCode")} - {m3u8_info.get("message")}'
                # 35 means media unavailable, but retries work
                if m3u8_info.get('errorCode') != 35 or attempt >= retries:
                    raise ExtractorError(last_error)
```

Reply:
Seems to work. I made some small changes.

Reply:
the point of `last_error` is to report warning when retrying. If you don't set it when errorCode=35, what is the point. Also, `'Couldn\'t retrieve m3u8 URL'` is not a descriptive error. Its better to give last_error as the message so user can have an idea what the issue is

my suggestion should have covered all these

Reply:
Yeah you're right. Fixed.